### PR TITLE
Repos in the dashboard

### PR DESCRIPTION
**Pull request by SaraVieira**
Adds repos to the dashboard
CleanShot 2020-06-30 at 15 45 40
CleanShot 2020-06-30 at 15 45 31
CleanShot 2020-06-30 at 15 46 37

### CODE REVIEW

**Comment by lbogdan** [comment]
Build for latest commit https://github.com/codesandbox/codesandbox-client/pull/4461/commits/ce55d23814b4e72c54c4180f6c4f6b48917a7bfc failed.

File: packages/app/src/app/overmind/actions.ts [hunk 1]
```diff
@@ -66,6 +66,7 @@ type ModalName =
   | 'deleteSandbox'
   | 'feedback'
   | 'forkServerModal'
+  | 'newSandbox'
```
Comment (CompuIves) [inline packages/app/src/app/overmind/actions.ts:69]: We have a new way to render the "New Sandbox" modal, which allows for props. I can show you with a later suggestion.

File: packages/app/src/app/overmind/effects/gql/dashboard/queries.ts [hunk 2]
```diff
@@ -127,6 +131,33 @@ export const getCollections: Query<
   ${sidebarCollectionDashboard}
 `;
 
+export const getRepos: Query<GetReposQuery, GetReposQueryVariables> = gql`
```
Comment (CompuIves) [inline packages/app/src/app/overmind/effects/gql/dashboard/queries.ts:134]: Would be good to call this `getPersonalRepos`, makes it more clear that there are two (also a team one).

File: packages/app/src/app/overmind/namespaces/dashboard/actions.ts [hunk 3]
```diff
@@ -398,6 +400,75 @@ export const getSandboxesByPath: AsyncAction<string> = async (
   }
 };
 
+export const getReposByPath: AsyncAction | Action<string> = async (
```
Comment (CompuIves) [inline packages/app/src/app/overmind/namespaces/dashboard/actions.ts:403]:
```suggestion
export const getReposByPath: AsyncAction<string> = async (
```

File: packages/app/src/app/overmind/namespaces/dashboard/actions.ts [hunk 4]
```diff
@@ -398,6 +400,75 @@ export const getSandboxesByPath: AsyncAction<string> = async (
...
+  { state, effects },
+  path
+  // eslint-disable-next-line consistent-return
+) => {
+  const { dashboard } = state;
```
Comment (CompuIves) [inline packages/app/src/app/overmind/namespaces/dashboard/actions.ts:408]: Nice implementation!

File: packages/app/src/app/overmind/namespaces/dashboard/utils.ts [hunk 5]
```diff
@@ -1,6 +1,10 @@
 import { SidebarCollectionDashboardFragment as Collection } from 'app/graphql/types';
 import { DELETE_ME_COLLECTION } from './state';
 
+export const isRepoPage = location.pathname.includes('/repositories');
```
Comment (CompuIves) [inline packages/app/src/app/overmind/namespaces/dashboard/utils.ts:4]: This doesn't work when you switch pages, and I think it's in general not a good practice to rely on URL string operations to determine what should be done. For example, if there's a directory in the dashboard called "repositories" this could break, and there are probably more examples where it doesn't work.

It would be better if we'd pass in the actions in overmind what the page is that we're dealing with, so we can base the logic on that instead.

File: packages/app/src/app/overmind/namespaces/dashboard/utils.ts [hunk 6]
```diff
@@ -1,6 +1,10 @@
...
+
+export const repoName = location.pathname.split('/repositories/')[1];
```
Comment (CompuIves) [inline packages/app/src/app/overmind/namespaces/dashboard/utils.ts:6]: Same for this, I would pass the `repoName` in to an action instead of gathering it from the URL.

File: packages/app/src/app/overmind/namespaces/dashboard/actions.ts [hunk 7]
```diff
@@ -571,14 +642,34 @@ export const deleteSandbox: AsyncAction<string[]> = async (
   const { user } = state;
   if (!user) return;
   const oldSandboxes = state.dashboard.sandboxes;
-  actions.dashboard.deleteSandboxFromState(ids);
+  if (!isRepoPage) {
```
Comment (CompuIves) [inline packages/app/src/app/overmind/namespaces/dashboard/actions.ts:645]: Would it maybe be better that we also delete sandboxes from `state.dashboard.sandboxes.REPOS` in the `actions.dashboard.deleteSandboxFromState`? If you delete a sandbox tied to the GH repo from Drafts, it should also not be visible from Repositories I reckon.

File: packages/app/src/app/overmind/namespaces/dashboard/actions.ts [hunk 8]
```diff
@@ -683,25 +774,70 @@ export const renameFolderInState: Action<{ path: string; newPath: string }> = (
...
   oldTitle: string;
 }> = async ({ effects, actions }, { id, title, oldTitle }) => {
-  actions.dashboard.renameSandboxInState({
-    id,
-    title,
-  });
-  try {
-    await effects.gql.mutations.renameSandbox({
+  if (isRepoPage) {
+    actions.dashboard.renameRepoSandboxInState({
```
Comment (CompuIves) [inline packages/app/src/app/overmind/namespaces/dashboard/actions.ts:813]: Same for here, shouldn't we handle renaming repo sandboxxes in `renameSandboxInState`. I'd like to know your opinion on this as well, @christianalfoni.

An idea that I had before would be that we actually keep all sandboxes in one array, and create `derived` functions for ordering them in `Recents`, `AllSandboxes`, `Repos`, etc... That would allow us to eg. only rename one instance of a sandbox and it's reflected everywhere immediately, also would save a lot of work when we're going to add live updates. I realize that this should be a PR on its own, but I'm thinking that if we're going to plan to do this anyway, then it would be better if we already put the renaming logic in a single function (`renameRepoSandboxInState`) instead of calling it conditionally.

File: packages/app/src/app/overmind/namespaces/dashboard/actions.ts [hunk 9]
```diff
@@ -776,13 +912,18 @@ export const makeTemplates: AsyncAction<{ sandboxIds: string[] }> = async (
   { sandboxIds: ids }
 ) => {
   const oldSandboxes = state.dashboard.sandboxes;
-  actions.dashboard.deleteSandboxFromState(ids);
+  if (!isRepoPage) {
```
Comment (CompuIves) [inline packages/app/src/app/overmind/namespaces/dashboard/actions.ts:915]: I would also make this conditional.

File: packages/app/src/app/overmind/namespaces/dashboard/actions.ts [hunk 10]
```diff
@@ -1026,14 +1167,51 @@ export const changeSandboxPrivacy: AsyncAction<{
...
+        }),
+      },
+    };
+  };
   // optimistic update
-  actions.dashboard.changeSandboxPrivacyInState({ id, privacy });
+  if (isRepoPage) {
+    repoChangePrivacy(privacy);
+  } else {
+    actions.dashboard.changeSandboxPrivacyInState({ id, privacy });
```
Comment (CompuIves) [inline packages/app/src/app/overmind/namespaces/dashboard/actions.ts:1200]: I would also put the repo privacy change in this function instead of 2 entry points.

File: packages/app/src/app/pages/NewDashboard/Components/Breadcrumbs/index.tsx [hunk 11]
```diff
@@ -1,43 +1,45 @@
...
-    >
-      All Sandboxes {path && ' / '}
-    </Link>
-    {path
-      ? path.split('/').map((p, i) => {
-          const partPath = path
-            .split('/')
-            .slice(0, i + 1)
-            .join('/');
+  const makeLink = (p: string, link: string) => {
```
Comment (CompuIves) [inline packages/app/src/app/pages/NewDashboard/Components/Breadcrumbs/index.tsx:8]: There was a bug in breadcrumbs for 2 folders deep, which is why I refactored this component. I think it would be better to base the changes in `Breadcrumbs` on the new version instead, I can definitely help with that!

File: packages/app/src/app/pages/NewDashboard/Components/EmptyScreen/index.tsx [hunk 12]
```diff
@@ -39,6 +40,29 @@ export const EmptyScreen: React.FC<EmptyScreenProps> = ({
     );
   }
 
+  if (page === 'repos') {
```
Comment (CompuIves) [inline packages/app/src/app/pages/NewDashboard/Components/EmptyScreen/index.tsx:43]: Nice!!

File: packages/app/src/app/pages/NewDashboard/Components/Folder/FolderCard.tsx [hunk 13]
```diff
@@ -102,7 +102,7 @@ export const FolderCard: React.FC<FolderItemComponentProps> = ({
         <IconButton
           name="more"
           size={9}
-          title="Sandbox actions"
+          title="Folder actions"
```
Comment (CompuIves) [inline packages/app/src/app/pages/NewDashboard/Components/Folder/FolderCard.tsx:105]: Good find!

File: packages/app/src/app/pages/NewDashboard/Components/Header/index.tsx [hunk 14]
```diff
@@ -52,7 +54,7 @@ export const Header = ({
           {title}
         </Text>
       ) : (
-        <Breadcrumbs activeTeam={activeTeam} path={path} />
+        <Breadcrumbs repos={repos} param={path} />
```
Comment (CompuIves) [inline packages/app/src/app/pages/NewDashboard/Components/Header/index.tsx:62]: `activeTeam` needs to be passed in the crumbs, because it's used to determine the URL (`teamId` is passed in the URL).

File: packages/app/src/app/pages/NewDashboard/Components/Repo/ImportRepo.tsx [hunk 15]
```diff
@@ -0,0 +1,34 @@
+import React from 'react';
+import { useOvermind } from 'app/overmind';
+import { Stack, Text, Button, Icon } from '@codesandbox/components';
+import css from '@styled-system/css';
+
+export const ImportRepo = () => {
+  const { actions } = useOvermind();
+
+  const onClick = () => actions.modalOpened({ modal: 'newSandbox' });
```
Comment (CompuIves) [inline packages/app/src/app/pages/NewDashboard/Components/Repo/ImportRepo.tsx:9]: There's a new one, I think it's 
```suggestion
  const onClick = () => actions.modals.newSandboxModal.open({});
```

File: packages/app/src/app/pages/NewDashboard/Components/Repo/RepoCard.tsx [hunk 16]
```diff
@@ -0,0 +1,86 @@
...
+            '-webkit-box-orient': 'vertical',
+          })}
+        >
+          {props.owner}/{name}
+        </Text>
+        <Text size={3} variant="muted" weight="medium">
+          {props.branch}
+        </Text>
+      </Stack>
+      <IconButton
```
Comment (CompuIves) [inline packages/app/src/app/pages/NewDashboard/Components/Repo/RepoCard.tsx:78]: `IconButton` is really expensive to render for some reason, it takes over 3ms to render at times, which makes the page super slow when scrolling. I have a trick on the `SandboxCard` component where we can prevent the slowdown by just rendering an icon when we're scrolling. It would be good if we could use that here as well!

File: packages/app/src/app/pages/NewDashboard/Components/Repo/RepoListItem.tsx [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+      <Column span={[0, 4, 4]} as={Stack} align="center">
+        <Text size={3} weight="medium" variant="muted">
+          {props.owner}
+        </Text>
+      </Column>
+      <Column span={[0, 3, 3]} as={Stack} align="center">
+        {/* empty column to align with sandbox list items */}
+      </Column>
+    </Grid>
+    <IconButton
```
Comment (CompuIves) [inline packages/app/src/app/pages/NewDashboard/Components/Repo/RepoListItem.tsx:72]: Same one here, would be good to render something different while scrolling. `SandboxCard` has an example for it.

File: packages/app/src/app/pages/NewDashboard/Components/Sandbox/index.tsx [hunk 18]
```diff
@@ -239,6 +239,8 @@ const GenericSandbox = ({ isScrolling, item }: GenericSandboxProps) => {
   };
 
   const sandboxProps = {
+    prNumber: item.sandbox.prNumber,
+    originalGit: item.sandbox.originalGit,
```
Comment (CompuIves) [inline packages/app/src/app/pages/NewDashboard/Components/Sandbox/index.tsx:243]: Aren't these props not on `sandbox` itself? (on line 249)

File: packages/app/src/app/pages/NewDashboard/Components/Selection/ContextMenu.tsx [hunk 19]
```diff
@@ -76,6 +100,10 @@ export const ContextMenu: React.FC<IContextMenuProps> = ({
     menu = <SandboxMenu item={selectedItems[0]} setRenaming={setRenaming} />;
   } else if (selectedItems[0].type === 'folder') {
     menu = <FolderMenu folder={selectedItems[0]} setRenaming={setRenaming} />;
+  } else if (selectedItems[0].type === 'repo') {
```
Comment (CompuIves) [inline packages/app/src/app/pages/NewDashboard/Components/Selection/ContextMenu.tsx:103]: Very nice!

File: packages/app/src/app/pages/NewDashboard/Components/Selection/ContextMenu.tsx [hunk 20]
```diff
@@ -342,6 +370,35 @@ const SandboxMenu: React.FC<SandboxMenuProps> = ({ item, setRenaming }) => {
   );
 };
 
+const RepoMenu = ({ repo }) => {
+  const { effects } = useOvermind();
+  const { visible, setVisibility, position } = React.useContext(Context);
+  const link = `https://github.com/${repo.owner}/${repo.name}/tree/${repo.branch}`;
+  const csbLink = `https://codesandbox.io/s/github/${repo.owner}/${repo.name}/tree/${repo.branch}`;
```
Comment (CompuIves) [inline packages/app/src/app/pages/NewDashboard/Components/Selection/ContextMenu.tsx:377]: For generating links it would be better to use the helper function. This wouldn't work for example for sandboxes that are linked to a commit. in `url-generators.ts` there is a `sandboxLink` (or similar name) function that handles the `repo` info.

File: packages/app/src/app/pages/NewDashboard/Components/Selection/ContextMenu.tsx [hunk 21]
```diff
@@ -41,16 +44,38 @@ export const ContextMenu: React.FC<IContextMenuProps> = ({
...
+    | DashboardRepo
+    | DashboardNewMasterBranch
   > = selectedIds.map(id => {
     if (id.startsWith('/')) {
+      if (repos.length) {
+        const repo = repos.find(f => '/' + f.name === id);
+        return { type: 'repo', ...repo };
+      }
+
+      if (id.includes('/github')) {
```
Comment (CompuIves) [inline packages/app/src/app/pages/NewDashboard/Components/Selection/ContextMenu.tsx:67]: Is there a more explicit way we can check here? To make sure nothing goes wrong? Like checking for `id.startsWith(` instead maybe?

File: packages/app/src/app/pages/NewDashboard/Components/Selection/ContextMenu.tsx [hunk 22]
```diff
@@ -496,6 +559,48 @@ const MultiMenu = ({ selectedItems }: IMultiMenuProps) => {
...
+    <Menu.ContextMenu
+      visible={visible}
+      setVisibility={setVisibility}
+      position={position}
+      style={{ width: 120 }}
+    >
+      <MenuItem
+        onSelect={() =>
+          window.open(
+            `https://codesandbox.io/s/github/${repo.owner}/${repo.name}`
```
Comment (CompuIves) [inline packages/app/src/app/pages/NewDashboard/Components/Selection/ContextMenu.tsx:584]: Same here, is there a better way that we can generate the URL?

Comment (CompuIves) [reply packages/app/src/app/pages/NewDashboard/Components/Selection/ContextMenu.tsx:584]: I think it would be better to use relative URLs as well (so it works on staging for example).

File: packages/app/src/app/pages/NewDashboard/Components/Selection/index.tsx [hunk 23]
```diff
@@ -562,9 +570,50 @@ export const SelectionProvider: React.FC<SelectionProviderProps> = ({
     else if (new Date().getTime() - callbackCalledAt.current > 60) callback();
   };
 
-  const onContainerMouseUp = event => {
+  const onContainerMouseUp = () => {
     if (drawingRect) resetSelectionRect();
   };
+  if (page === 'repos') {
```
Comment (CompuIves) [inline packages/app/src/app/pages/NewDashboard/Components/Selection/index.tsx:576]: Do we need a completely different one for the `'repos'` page?

File: packages/app/src/app/pages/NewDashboard/Components/VariableGrid/index.tsx [hunk 24]
```diff
@@ -71,8 +80,12 @@ const ComponentForTypes: ComponentForTypes = {
     <Sandbox item={props.item} isScrolling={props.isScrolling} />
   )),
   folder: props => <Folder {...props.item} />,
+  repo: props => <Repo {...props.item} />,
+  // @ts-ignore TODO: find a better way to type this
```
Comment (CompuIves) [inline packages/app/src/app/pages/NewDashboard/Components/VariableGrid/index.tsx:84]:
```suggestion
```

File: packages/app/src/app/pages/NewDashboard/Content/routes/Repositories/index.tsx [hunk 25]
```diff
@@ -0,0 +1,98 @@
...
+  React.useEffect(() => {
+    const p = home ? null : param;
+    actions.dashboard.getReposByPath(p);
+  }, [param, actions.dashboard, activeTeam, home]);
+
+  const activeSandboxes =
+    (sandboxes.REPOS && Object.values(sandboxes.REPOS)) || [];
+
+  const itemsToShow = (): DashboardGridItem[] => {
+    if (!sandboxes.REPOS) {
```
Comment (CompuIves) [inline packages/app/src/app/pages/NewDashboard/Content/routes/Repositories/index.tsx:34]: We need to check explicitly with `== null` here, because it now will also show the skeleton in case `sandboxes.REPOS = []` I believe.

File: packages/app/src/app/pages/NewDashboard/Content/routes/Repositories/index.tsx [hunk 26]
```diff
@@ -0,0 +1,98 @@
...
+          ]
+        : items;
+    }
+
+    return [{ type: 'skeleton-row' }, { type: 'skeleton-row' }];
+  };
+
+  const possibleTemplates = itemsToShow()
+    // @ts-ignore
+    .filter(s => s.sandbox)
```
Comment (CompuIves) [inline packages/app/src/app/pages/NewDashboard/Content/routes/Repositories/index.tsx:62]: Hmm, why do we do this? And what is TS complaining about?

File: packages/app/src/app/pages/NewDashboard/Content/routes/Repositories/index.tsx [hunk 27]
```diff
@@ -0,0 +1,98 @@
...
+    },
+  } = useOvermind();
+
+  React.useEffect(() => {
+    const p = home ? null : param;
+    actions.dashboard.getReposByPath(p);
+  }, [param, actions.dashboard, activeTeam, home]);
+
+  const activeSandboxes =
+    (sandboxes.REPOS && Object.values(sandboxes.REPOS)) || [];
```
Comment (CompuIves) [inline packages/app/src/app/pages/NewDashboard/Content/routes/Repositories/index.tsx:31]: `Object.values` is quite expensive. It maps over all the values and puts them in an array. Maybe we could check explicitly for `null` instead here?

File: packages/app/src/app/pages/NewDashboard/Content/routes/Repositories/useFilteredItems.ts [hunk 28]
```diff
@@ -0,0 +1,60 @@
...
+      },
+    },
+  } = useOvermind();
+  const [items, setItems] = useState<Array<DashboardGridItem>>([]);
+  const sandboxesForPath =
+    sandboxes.REPOS && param
+      ? getFilteredSandboxes(sandboxes.REPOS[param].sandboxes || [])
+      : [];
+
+  const repos = (sandboxes.REPOS && Object.values(sandboxes.REPOS)) || [];
```
Comment (CompuIves) [inline packages/app/src/app/pages/NewDashboard/Content/routes/Repositories/useFilteredItems.ts:28]: Same here, I think it's better to not use `Object.values` but rather something like `== null`.

File: packages/app/src/app/pages/NewDashboard/Sidebar/index.tsx [hunk 29]
```diff
@@ -172,6 +172,13 @@ export const Sidebar: React.FC<SidebarProps> = ({
             ]}
           />
 
+          <RowItem
+            name="Repositories"
+            page="repos"
+            path={dashboardUrls.repos(activeTeam)}
```
Comment (CompuIves) [inline packages/app/src/app/pages/NewDashboard/Sidebar/index.tsx:178]: This is super nice!

File: packages/app/src/app/pages/NewDashboard/types.ts [hunk 30]
```diff
@@ -12,14 +13,20 @@ export type DashboardBaseFolder = {
 
 export type DashboardSandbox = {
   type: 'sandbox';
-  sandbox: SandboxFragmentDashboardFragment;
+  sandbox: SandboxFragmentDashboardFragment & {
```
Comment (CompuIves) [inline packages/app/src/app/pages/NewDashboard/types.ts:16]: Maybe we should add the properties to the `SandboxFragmentDashboardFragment` instead? Since it seems like we're going to use it in other pages besides `Repositories` as well (from what I've seen).

File: packages/common/src/global.css [hunk 31]
```diff
@@ -1,6 +1,6 @@
 html,
 body {
-  font-family: 'Roboto', sans-serif;
+  font-family: 'Inter', sans-serif;
```
Comment (CompuIves) [inline packages/common/src/global.css:3]: This might break some styling in other places, we definitely need to test all pages to make sure that everything still works with this font.

Alternatively, I think we should do this in a separate PR. Then we are not bringing this PR out of scope.

Comment (SaraVieira) [reply packages/common/src/global.css:3]: removed, it was a request from danny

File: packages/app/src/app/pages/NewDashboard/types.ts (hunk 30 above)
Comment (SaraVieira) [reply packages/app/src/app/pages/NewDashboard/types.ts:16]: hm?

File: packages/app/src/app/pages/NewDashboard/Content/routes/Repositories/useFilteredItems.ts (hunk 28 above)
Comment (SaraVieira) [reply packages/app/src/app/pages/NewDashboard/Content/routes/Repositories/useFilteredItems.ts:28]: fixed

Comment (SaraVieira) [reply packages/app/src/app/pages/NewDashboard/Content/routes/Repositories/useFilteredItems.ts:28]: Ohhh, no because I actually want the values of each key in the object

File: packages/app/src/app/pages/NewDashboard/Content/routes/Repositories/index.tsx (hunk 27 above)
Comment (SaraVieira) [reply packages/app/src/app/pages/NewDashboard/Content/routes/Repositories/index.tsx:31]: That's the thing I need that, to map the values and put them in an array ahah

File: packages/app/src/app/overmind/namespaces/dashboard/utils.ts (hunk 5 above)
Comment (SaraVieira) [reply packages/app/src/app/overmind/namespaces/dashboard/utils.ts:4]: All done :)

File: packages/app/src/app/overmind/namespaces/dashboard/actions.ts (hunk 3 above)
Comment (SaraVieira) [reply packages/app/src/app/overmind/namespaces/dashboard/actions.ts:413]: because of line 421 It can also be a normal actions

File: packages/app/src/app/overmind/effects/gql/dashboard/queries.ts (hunk 2 above)
Comment (SaraVieira) [reply packages/app/src/app/overmind/effects/gql/dashboard/queries.ts]: good point! Done :)

File: packages/app/src/app/pages/NewDashboard/Components/Selection/ContextMenu.tsx (hunk 21 above)
Comment (SaraVieira) [reply packages/app/src/app/pages/NewDashboard/Components/Selection/ContextMenu.tsx:70]: changed

File: packages/app/src/app/pages/NewDashboard/Components/Selection/index.tsx (hunk 23 above)
Comment (SaraVieira) [reply packages/app/src/app/pages/NewDashboard/Components/Selection/index.tsx:576]: omg yes, there is noe fancy stuff, I don't think the contextMenu should be there tbh

File: packages/app/src/app/pages/NewDashboard/Components/VariableGrid/index.tsx (hunk 24 above)
Comment (SaraVieira) [reply packages/app/src/app/pages/NewDashboard/Components/VariableGrid/index.tsx:94]: Removed :)

File: packages/app/src/app/pages/NewDashboard/Content/routes/Repositories/index.tsx (hunk 25 above)
Comment (SaraVieira) [reply packages/app/src/app/pages/NewDashboard/Content/routes/Repositories/index.tsx:34]: changed :)

File: packages/app/src/app/pages/NewDashboard/Content/routes/Repositories/index.tsx (hunk 26 above)
Comment (SaraVieira) [reply packages/app/src/app/pages/NewDashboard/Content/routes/Repositories/index.tsx:62]: it says there is no sandbox in the type 🤷‍♀️

File: packages/app/src/app/pages/NewDashboard/Components/Selection/ContextMenu.tsx (hunk 20 above)
Comment (SaraVieira) [reply packages/app/src/app/pages/NewDashboard/Components/Selection/ContextMenu.tsx:397]: Done :)

File: packages/app/src/app/pages/NewDashboard/Components/Selection/ContextMenu.tsx (hunk 22 above)
Comment (SaraVieira) [reply packages/app/src/app/pages/NewDashboard/Components/Selection/ContextMenu.tsx]: done :)

File: packages/app/src/app/overmind/namespaces/dashboard/utils.ts (hunk 6 above)
Comment (SaraVieira) [reply packages/app/src/app/overmind/namespaces/dashboard/utils.ts:4]: I honestly think its better to get it from the URL, in the cases where we are in the repo page this will never break and we already send the page

I think this makes it easier to not screw up

File: packages/app/src/app/pages/NewDashboard/Content/routes/Repositories/index.tsx (hunk 27 above)
Comment (CompuIves) [reply packages/app/src/app/pages/NewDashboard/Content/routes/Repositories/index.tsx:31]: Ahhh ohhh, yep this makes sense. I initially thought you were only using it to find whether the array was empty or not. Sorry! 😂

File: packages/app/src/app/overmind/namespaces/dashboard/actions.ts (hunk 9 above)
Comment (SaraVieira) [reply packages/app/src/app/overmind/namespaces/dashboard/actions.ts]: Moved the condition to the function itself

File: packages/app/src/app/overmind/namespaces/dashboard/actions.ts (hunk 10 above)
Comment (SaraVieira) [reply packages/app/src/app/overmind/namespaces/dashboard/actions.ts:1220]: done

File: packages/app/src/app/pages/NewDashboard/Components/Header/index.tsx (hunk 14 above)
Comment (SaraVieira) [reply packages/app/src/app/pages/NewDashboard/Components/Header/index.tsx:62]: done

File: packages/app/src/app/pages/NewDashboard/Components/Breadcrumbs/index.tsx (hunk 11 above)
Comment (SaraVieira) [reply packages/app/src/app/pages/NewDashboard/Components/Breadcrumbs/index.tsx:8]: updated :)

File: packages/app/src/app/pages/NewDashboard/Components/Repo/ImportRepo.tsx (hunk 15 above)
Comment (SaraVieira) [reply packages/app/src/app/pages/NewDashboard/Components/Repo/ImportRepo.tsx:9]: updated

File: packages/app/src/app/overmind/actions.ts (hunk 1 above)
Comment (SaraVieira) [reply packages/app/src/app/overmind/actions.ts:69]: removed :)

File: packages/app/src/app/pages/NewDashboard/Components/Repo/RepoCard.tsx (hunk 16 above)
Comment (SaraVieira) [reply packages/app/src/app/pages/NewDashboard/Components/Repo/RepoCard.tsx]: done :)

File: packages/app/src/app/pages/NewDashboard/Components/Repo/RepoListItem.tsx (hunk 17 above)
Comment (SaraVieira) [reply packages/app/src/app/pages/NewDashboard/Components/Repo/RepoListItem.tsx]: done

File: packages/app/src/app/overmind/namespaces/dashboard/actions.ts (hunk 3 above)
Comment (SaraVieira) [reply packages/app/src/app/overmind/namespaces/dashboard/actions.ts:413]: nvm, done

File: packages/app/src/app/overmind/namespaces/dashboard/actions.ts (hunk 7 above)
Comment (SaraVieira) [reply packages/app/src/app/overmind/namespaces/dashboard/actions.ts]: done

File: packages/app/src/app/overmind/namespaces/dashboard/actions.ts (hunk 8 above)
Comment (SaraVieira) [reply packages/app/src/app/overmind/namespaces/dashboard/actions.ts:829]: this is done, I don't know what the rest is ahah

File: packages/app/src/app/pages/NewDashboard/Components/Sandbox/index.tsx (hunk 18 above)
Comment (SaraVieira) [reply packages/app/src/app/pages/NewDashboard/Components/Sandbox/index.tsx:245]: But the sandbox isn't passed in totality down

Comment (CompuIves) [reply packages/app/src/app/pages/NewDashboard/Components/Sandbox/index.tsx:245]: It is, on line 251 the `item.sandbox` is passed down in totality. Or is there another reason to send down these props separately?

File: packages/app/src/app/pages/NewDashboard/Components/Selection/ContextMenu.tsx [hunk 32]
```diff
@@ -41,16 +47,39 @@ export const ContextMenu: React.FC<IContextMenuProps> = ({
...
   const selectedItems: Array<
-    DashboardFolder | DashboardSandbox | DashboardTemplate
+    | DashboardFolder
+    | DashboardSandbox
+    | DashboardTemplate
+    | DashboardRepo
+    | DashboardNewMasterBranch
   > = selectedIds.map(id => {
     if (id.startsWith('/')) {
+      if (repos.length) {
```
Comment (CompuIves) [inline packages/app/src/app/pages/NewDashboard/Components/Selection/ContextMenu.tsx:66]: Should we maybe change `repos?: ` to `repos: ` in the props? Since we're assuming here that `repos` cannot be `undefined`.

File: packages/app/src/app/pages/NewDashboard/Components/Selection/index.tsx (hunk 23 above)
Comment (CompuIves) [reply packages/app/src/app/pages/NewDashboard/Components/Selection/index.tsx:576]: Do we still need it here in that case?

File: packages/app/src/app/pages/NewDashboard/types.ts (hunk 30 above)
Comment (CompuIves) [reply packages/app/src/app/pages/NewDashboard/types.ts:16]: I'm wondering, do we also show specific properties or menu items for repo sandboxes in other places than the repo page? If that's the case we could also change the fragment query of sandbox itself to include these properties.

File: packages/common/src/global.css [hunk 33]
```diff
@@ -6,14 +6,14 @@ body {
   -moz-osx-font-smoothing: grayscale;
   font-smoothing: auto;
   text-rendering: optimizeLegibility;
-  font-smooth: always;
+  font-smooth: alays;
```
Comment (CompuIves) [inline packages/common/src/global.css:9]:
```suggestion
  font-smooth: always;
```

File: packages/app/src/app/overmind/namespaces/dashboard/actions.ts [hunk 34]
```diff
@@ -6,10 +7,21 @@ import {
...
   TemplateFragmentDashboardFragment,
   SandboxFragmentDashboardFragment,
+  RepoFragmentDashboardFragment,
 } from 'app/graphql/types';
 import { TEAM_ID_LOCAL_STORAGE } from 'app/overmind/utils/team';
 import { OrderBy, sandboxesTypes } from './state';
-import { getDecoratedCollection } from './utilts';
+import { getDecoratedCollection, repoName } from './utils';
+
+type PageTypes =
```
[the remark this answers is not in the document]
Comment (SaraVieira) [reply packages/app/src/app/overmind/namespaces/dashboard/actions.ts:16]: done, moved a bunch of stuff there

File: packages/app/src/app/overmind/namespaces/dashboard/utils.ts (hunk 6 above)
Comment (SaraVieira) [reply packages/app/src/app/overmind/namespaces/dashboard/utils.ts:4]: this does work when you move pages, I tested this a lot to make sure it wouldn't break but sure I can pass the repo name into the action

Comment (SaraVieira) [reply packages/app/src/app/overmind/namespaces/dashboard/utils.ts]: done

**Comment by christianalfoni** [comment]
Hi there! 😄 

Okay, first of all! This is an insanely complicated implementation where @CompuIves has a TON of existing knowledge having implemented the dashboard once before. So what I think is important to remember here is that the goal was not only to implement a new dashboard, but also transfer the ownership from @CompuIves. It is really great we get this review and get some laste minute knowledge sharing and improvements, but also... this is not the last time we will touch the dashboard 😄  

So I am super pleased that the code is now also manageable for me personally, as it is consistent with our other implementations as well. And browsing through the comments and fixes here, I believe we are in a shipable state after the last ones are resolved. And also in a really good state to fix any issues popping up, and also continue extending its features, as we are now 3 people who feels comfortable changing the code, instead of just one 😄 

I do have some suggestions on how we can structure things a little bit differently to improve consistency and reduce some boilerplate code, but that would be for the next time we look at the dashboard. In advance, sorry if any of these suggestions are misleading as there are certainly parts of the implementation I have not grasped skimming through code 😄 : 

1. I see that Sandboxes are now organized explicitly by page and thinking about this, I think it makes total sense. Though... instead of storing the data for each page, I would rather split storing data and deriving what data to show for each page:

```ts
{
  sandboxes: { [id: string]: Sandbox },
  templates: { [id: string]: Template },
  repos: { [path: string]: Repo }, // This is the full type, including `sandboxes`
  currentPage: Page
  currentSandboxes: {
    [Page.DRAFTS]: Sandbox[], // derive sandboxes based on page
    [Page.TEMPLATES]: Template[], // derive templates based on page
    ...
  }
}
```

This will also remove things like `recentSandboxesByTime` and `deletedSandboxesByTime`, as those are now the derived for the specific page. It would make all the derived dirty when moving between pages, but since you can move sandboxes between pages it makes sense to recompute them when you open a page as things could have changed. Also whenever a page is opened a new query should run and update the data source, ex `sandboxes`, if the ID is not already there. 

Anyways... this ensures that the UI is always consistent by recomputing, instead of manually updating the data sources.

What this also now allows is having a single action for opening a page:

```ts
export const openPage = ({ state, effects }, page) => {
  state.dashboard.currentPage = page
  try {
    switch (page) {
      case Page.DRAFTS: {
        ... fetch some data
      }
      case Page.RECENT: {
        ... fetch some data
      }
    }
  } catch () {}
}
```

This also simplifies things like deleting:

```ts
export const deleteSandboxes: AsyncAction<{
  ids: string[];
> = async ({ state, effects, actions }, { ids, page }) => {
  const { user } = state;
  if (!user) return;
  
  const sandboxesToDelete = ids.reduce((id) => aggr.concat(state.dashboard.sandboxes[id]), [])

  sandboxesToDelete.forEach(sandbox => {
    delete state.dashboard.sandboxes[sandbox.id]
  }) 

  try {
    await effects.gql.mutations.deleteSandboxes({
      sandboxIds: ids,
    });
  } catch (error) {
    sandboxesToDelete.forEach((sandbox) => {
      state.dashboard.sandboxes[sandbox.id] = sandbox
    })

    effects.notificationToast.error(
      'There was a problem deleting your Sandbox'
    );
  }
};
```

Okay, those where the things I think could clean up a little bit and also ensure some more consistency. 

Anyways, awesome work getting this in and also hammering through feedback... and types... just one more run and it is shipable! 😄 ❤️

File: packages/app/src/app/pages/NewDashboard/Components/Selection/ContextMenu.tsx (hunk 32 above)
Comment (SaraVieira) [reply packages/app/src/app/pages/NewDashboard/Components/Selection/ContextMenu.tsx:66]: my bad, also checked if repos exist now

File: packages/app/src/app/pages/NewDashboard/Components/Sandbox/index.tsx (hunk 18 above)
Comment (SaraVieira) [reply packages/app/src/app/pages/NewDashboard/Components/Sandbox/index.tsx:254]: ohh, my bad, used that instead

File: packages/app/src/app/pages/NewDashboard/Components/Selection/index.tsx (hunk 23 above)
Comment (SaraVieira) [reply packages/app/src/app/pages/NewDashboard/Components/Selection/index.tsx:576]: spoke in slack

File: packages/app/src/app/pages/NewDashboard/types.ts (hunk 30 above)
Comment (SaraVieira) [reply packages/app/src/app/pages/NewDashboard/types.ts:19]: No we won't but that's the issue, if I try to get them TS will complain even thought I know they will exist since I am on the repos page

**Comment by CompuIves** [comment]
Nice!!! :shipit:
